### PR TITLE
build: refactor cmake

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
Signed-off-by: Lloyd-Pottiger <yan1579196623@gmail.com>

### What problem does this PR solve?

Issue Number: ref #6201

Problem Summary:

### What is changed and how it works?
1. test cpu feature like avx2, avx512 support during compiling
2. support ccache4.0+
3. replace `APPLE` with `ARCH_DARWIN`
4. replace `CMAKE_CXX_COMPILER_ID STREQUAL "Clang"` with `COMPILER_CLANG`
5. add `COMPILER_APPLE_CLANG ` for better distinguish between apple clang and llvm clang
6. use `ld` with `AppleClang` and `lld` with llvm `Clang`
7. separate some logic into a separate file

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by Lloyd-Pottiger** [comment]
compiler with `-mavx512` will success in platform without avx512 support, so ci build fail. I will try to find another way.

File: cmake/tools.cmake [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+message (STATUS "Using compiler:\n${COMPILER_SELF_IDENTIFICATION}")
+
+# Require minimum compiler versions
+set (CLANG_MINIMUM_VERSION 5)
+set (GCC_MINIMUM_VERSION 7)
+
+if (COMPILER_GCC)
+    if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS ${CLANG_MINIMUM_VERSION})
+        message (FATAL_ERROR "Compilation with GCC version ${CMAKE_CXX_COMPILER_VERSION} is not supported, minimum required version is ${CLANG_MINIMUM_VERSION}")
+    endif ()
```
Comment (JaySon-Huang) [inline cmake/tools.cmake:42]:
```suggestion
if (COMPILER_GCC)
    if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS ${GCC_MINIMUM_VERSION})
        message (FATAL_ERROR "Compilation with GCC version ${CMAKE_CXX_COMPILER_VERSION} is not supported, minimum required version is ${GCC_MINIMUM_VERSION}")
    endif ()
```

File: cmake/tools.cmake [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    set (COMPILER_GCC 1)
+    set(MOST_DEBUGGABLE_LEVEL -Og)
+elseif (CMAKE_CXX_COMPILER_ID MATCHES "AppleClang")
+    set (COMPILER_APPLE_CLANG 1)
+    set (COMPILER_CLANG 1) # Safe to treat AppleClang as a regular Clang, in general.
+    set(MOST_DEBUGGABLE_LEVEL -O0)
+elseif (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+    set (COMPILER_CLANG 1)
+    set(MOST_DEBUGGABLE_LEVEL -O0)
+else ()
```
Comment (JaySon-Huang) [inline cmake/tools.cmake:27]:
```suggestion
if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
    set (COMPILER_GCC 1)
    set (MOST_DEBUGGABLE_LEVEL -Og)
elseif (CMAKE_CXX_COMPILER_ID MATCHES "AppleClang")
    set (COMPILER_APPLE_CLANG 1)
    set (COMPILER_CLANG 1) # Safe to treat AppleClang as a regular Clang, in general.
    set (MOST_DEBUGGABLE_LEVEL -O0)
elseif (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
    set (COMPILER_CLANG 1)
    set (MOST_DEBUGGABLE_LEVEL -O0)
else ()
```

File: cmake/tools.cmake [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    set (CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fuse-ld=${LINKER_NAME}")
+    message(STATUS "Using linker: ${LINKER_NAME}")
+else()
+    message(STATUS "Using linker: <default>")
+endif()
+
+# Archiver
+
+if (COMPILER_GCC)
+    find_program (LLVM_AR_PATH NAMES "llvm-ar" "llvm-ar-15" "llvm-ar-14" "llvm-ar-13" "llvm-ar-12")
```
Comment (JaySon-Huang) [inline cmake/tools.cmake:84]: why gcc use `llvm-xxx`?

File: cmake/tools.cmake [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    message(STATUS "Using archiver: ${CMAKE_AR}")
+else ()
+    message (FATAL_ERROR "Cannot find ar.")
+endif ()
+
+
+# Ranlib
+
+if (COMPILER_GCC)
+    find_program (LLVM_RANLIB_PATH NAMES "llvm-ranlib" "llvm-ranlib-15" "llvm-ranlib-14" "llvm-ranlib-13" "llvm-ranlib-12")
```
Comment (JaySon-Huang) [inline cmake/tools.cmake:100]: ditto

File: cmake/tools.cmake [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    message(STATUS "Using ranlib: ${CMAKE_RANLIB}")
+else ()
+    message (FATAL_ERROR "Cannot find ranlib.")
+endif ()
+
+
+# Objcopy
+
+if (COMPILER_GCC)
+    find_program (OBJCOPY_PATH NAMES "llvm-objcopy" "llvm-objcopy-15" "llvm-objcopy-14" "llvm-objcopy-13" "llvm-objcopy-12" "objcopy")
```
Comment (JaySon-Huang) [inline cmake/tools.cmake:116]: ditto

File: cmake/tools.cmake (hunk 3 above)
Comment (Lloyd-Pottiger) [reply cmake/tools.cmake:84]: prefer to use llvm as backend, if there are no llvm toolchain, will use gcc-ar default.

**Review by solotzg — APPROVED** [review]
lgtm

**Comment by solotzg** [comment]
Can we just remove GCC support ?

File: cmake/tools.cmake (hunk 3 above)
Comment (JaySon-Huang) [reply cmake/tools.cmake:84]: seems there is no gcc-ar?

Comment (JaySon-Huang) [reply cmake/tools.cmake:84]: And can gcc work with llvm-ar?

Comment (Lloyd-Pottiger) [reply cmake/tools.cmake:84]: [gcc-ar](https://manpages.ubuntu.com/manpages/bionic/man1/aarch64-linux-gnu-gcc-ar.1.html) a wrapper around ar adding the --plugin option

**Comment by Lloyd-Pottiger** [comment]
@solotzg @JaySon-Huang add a warning message when using gcc.

File: cmake/tools.cmake (hunk 3 above)
Comment (Lloyd-Pottiger) [reply cmake/tools.cmake:85]: > And can gcc work with llvm-ar?

In theory, but now, using gcc to compile is not supported. 😆

File: cmake/target.cmake [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+if (CMAKE_SYSTEM MATCHES "Linux")
+    set (ARCH_LINUX 1)
+elseif (CMAKE_SYSTEM MATCHES "FreeBSD")
+    set (ARCH_FREEBSD 1)
+elseif (CMAKE_SYSTEM_NAME MATCHES "Darwin")
+    set (ARCH_DARWIN 1)
+endif ()
```
Comment (JaySon-Huang) [inline cmake/target.cmake:21]: maybe OS_LINUX/OS_FREEBSD/OS_DARWIN is more suitable to make it distinguish from `ARCH_AMD64`/`ARCH_ARM`

**Review by JaySon-Huang — APPROVED** [review]
LGTM with minor comment

**Pull request merged**